### DISCUSSION TITLE
Replacing `wit/deps/`

### DISCUSSION THREAD

**Discussion by lann** [bytecodealliance/wasm-pkg-tools#84] [General]
With the [addition of "nested packages"](https://github.com/bytecodealliance/wasm-tools/pull/1577) it is now possible to represent multiple packages in a single WIT file. We'd like to use this feature to set a new convention for the on-disk structure of WIT packages that are managed by `wasm-pkg-tools`:

- As a prerequisite, [a "package lock" file](https://github.com/bytecodealliance/wasm-pkg-tools/issues/15) will be introduced which will represent all of a project's (transitive) WIT package dependencies, e.g. `wkg.lock` (exact path TBD)
- The "locked" dependencies will be fetched and merged into a single `wit/resolved.wit` file (name to be bikeshod) with a boilerplate comment at the top e.g. "this file is generated; edits will be overwritten, etc"
- The `resolved.wit` file should be "deterministicishly" generated from the `wit.lock`, but will still likely be committed to source control in many cases to avoid typical registry auth / CI tooling issues. To help with the potential synchronization issue this can cause, there should be some kind of validation command e.g. `wkg wit pull --check` (?) to validate that the `resolved.wit` is up-to-date with the `wit.lock`.

**Comment by yoshuawuyts** [comment]
I'm curious if you could further elaborate why we need to check in a generated wit file if we also have a lock file? From what I gather it sounds like under this design each dependency will be referenced three times in each project:

- once by-version when imported
- once more in the lock file
- a third time in the generated deps file

In Node.js we used to vendor deps before lock files were added. But that basically stopped being necessary after lock files were added. So I guess I'm wondering what is different in this model that we would need both a lock file and vendoring?

**Reply by lann** [reply]
You wouldn't _have_ to, but it would help to avoid issues with tooling, config, auth etc. across development teams and CI.

**Reply by yoshuawuyts** [reply]
I feel like there is an implicit assumption here about the model we should go for and how we expect people will want to interact with WIT files in their projects. "Replacing `wit/deps`" could take many forms. For example:
- Docker stores its layers in a central storage, and users never get access to it. The closest you can get is to `oras cp` a remote url. I don't necessarily love that model, but it's certainly an option.
- npm uses lockfiles and generates a `node_modules` dir. But that dir is typically not checked in. The files in `node_modules` end up being symlinks to a shared, local dependency storage.
- Rust pulls down copies of all deps locally to a `target/` dir, not a central location. This directory is gitignored by default because people are expected to use `Cargo.lock` instead. There is also an additional `cargo vendor` subcommand for the cases when people don't.

I don't have any strong opinions right now on any particular direction we should take, but I do want to better understand why we believe this direction in particular is better than any of the alternatives.

To clarify why this has piqued my interest: I feel some unease about having a `wit/` directory which combines both user-authored files and auto-generated files. On a surface level it feels like those shouldn't be mixed. I assume I might be missing something in the motivation or direction that explains why that's a good thing, and so I'm trying to understand the broader direction I think?

**Reply by lann** [reply]
This isn't intended to represent an ideal end-state, just trying to acknowledge the interim reality of loosely coupled and ad-hoc toolchains we have today.

**Reply by yoshuawuyts** [reply]
Okay, so what you're saying is that this is basically just a formalization of how `wit-deps` already works today - but probably works a little better since we now have nested packages? Am I reading that right?

**Reply by macovedj** [reply]
In the Packaging SIG yesterday there was a bit of a discussion about how anything that can be represented by a deps folder can now be represented with nested packages.  There are also people who find the deps folder difficult to work with, and like the idea of everything being in a single file.  I don't think a final decision was made, but it was certainly considered that potentially we should recommend that various tools transition at whatever rate makes sense to prefer using nested packages over a deps folder.  Hope that context helps!

**Reply by lann** [reply]
I should also add that this proposal is only about generic wit tooling (i.e. `wkg wit` or wherever that ends up living). Language-specific tools like cargo-component or jco may choose to follow this convention or may choose to do something completely different.

In contrast, we _do_ want to standardize the "package lock" file across toolchains.

**Reply by lukewagner** [reply]
Agreed that it's not clear whether the generated WIT file should or shouldn't be checked in by convention (I can think of arguments in both directions.)

But just to check if the following aspects of the proposal are uncontroversial:
* Once there has been a depsolve that produces a lock file and the WIT package binaries have been fetched, a single-file WIT is generated (deterministically) from these package binaries as the next step of the overall build pipeline.
* This generated WIT file takes the place of the `deps/` directory.  Even if the WIT file is not checked in, it is generated and stored *somewhere* in the local project directory in a predictable location so that the developer can find it to see the "ground truth" of the world they are compiling against.
* This generated WIT file, an output directory, and a target `world` name (which itself could be optional if we add [unnamed worlds](https://github.com/WebAssembly/component-model/issues/313#issuecomment-1969987162)) are the sole inputs to `wit-bindgen` and thus `wit-bindgen` doesn't need to know anything about registries, directory structure, the current directory, `wit/`, or `deps/`.

Does anything sound off about any of that?

**Reply by lann** [reply]
> a single-file WIT is generated (deterministically) from these package binaries

~~"..._and any locally-managed WITs_"? That would seem to be necessary if "This generated WIT file [is one of the] sole inputs to `wit-bindgen`".~~

~~Edit: Which might suggest that `wit/resolved.wit` is a bad place to drop that file, if the "locally-managed WITs" would be `wit/*.wit`...~~

See other thread: https://github.com/bytecodealliance/wasm-pkg-tools/discussions/84#discussioncomment-10561549

**Comment by lukewagner** [comment]
One thing that wasn't touched on in the above proposal is the relation between the generated WIT file and any user-authored WIT.  My understanding is that there are two cases:
* The developer is authoring a component that targets a world already published somewhere (e.g. `wasi:http/proxy`) and so there is no user-authored WIT (just the registry-resolved name of the world).
* The developer authors some WIT that is *not* published to a registry but is instead committed along with the source code (and thus the name of this user-authored `world` is irrelevant, making it a good use case for [unnamed worlds](https://github.com/WebAssembly/component-model/issues/313#issuecomment-1969987162)).  However, this user-authored WIT *may* refer to registry-published interfaces or world names (via `import`, `export`, `use` or `include`).

In the former case, I was imagining there is a `wkg wit` command that takes a world name and then spits out the generated single-file WIT file containing the named world as a "root" from which all transitive dependencies are found and included.

In the latter case, I imagine that there is a different `wkg wit` command that takes the filename of the user-authored world and a (one-day optional) world name and then spits out the generated single-file WIT file containing the user-authored world as the "root" from which all transitive dependencies are found and included.

Does that sound right?

**Reply by lann** [reply]
I can see the generic workflow having two separate steps:

- Any time you update dependencies, you regenerate the lock file and `wit/resolved.wit`
  - This could happen explicitly (`wkg wit resolve`?) or implicitly (`wkg wit add <dependency>`?)
- In order to produce the actual input to wit-bindgen et al, you would further bundle everything in `wit/*.wit`, which would include that `resolved.wit` and/or user-authored WIT(s).
  - It might make sense to keep this simpler "glob a single directory" bundling step in `wit-bindgen` itself

**Reply by lukewagner** [reply]
Oh, I see.  So is the idea that, if you have a user-defined world, you end up with two WIT files: the one you wrote and the generated one with all your WIT dependencies and then you feed them both into wit-bindgen?
